File: GameLibrary/src/cc/lib/zombicide/anims/SonarAnimation.java
package cc.lib.zombicide.anims;

import cc.lib.game.AGraphics;
import cc.lib.game.GColor;
import cc.lib.math.Vector2D;
import cc.lib.zombicide.ZActor;
import cc.lib.zombicide.ZActorAnimation;
import cc.lib.zombicide.ZBoard;

/**
 * Animate concentric arcs that have a start and end point
 */
public class SonarAnimation extends ZActorAnimation {

    Vector2D start, end, dv;
    final int numArcs;
    final float startAngle;
    final float sweepAngle;
    final float radius;

    public SonarAnimation(ZActor actor, ZBoard board, int targetZone) {
        this(actor, actor.getRect().getCenter(), board.getZone(targetZone).getCenter(), 5, 20);
    }

    public SonarAnimation(ZActor actor, Vector2D start, Vector2D end, int numArcs, float sweepAngle) {
        super(actor, 2000);
        this.start = start;
        this.end = end;
        this.numArcs = numArcs;
        this.sweepAngle = sweepAngle;
        this.radius = end.sub(start).mag();
        this.startAngle = end.sub(start).angleOf() - sweepAngle/2;
        this.dv = end.sub(start).scaledBy(1f / numArcs);
    }

    @Override
    protected void draw(AGraphics g, float position, float dt) {
        g.setColor(GColor.WHITE);
        g.setLineWidth(3);
        float radiusStep = radius / numArcs;
        if (position <= .5f) {
            // draw the arcs emanating from the start
            int numArcsToDraw = Math.round(position * 2 * numArcs);
            //g.drawFilledCircle(start, radius/10);
            g.drawArc(start, radius/10, startAngle, sweepAngle);
            float r = radiusStep;
            for (int i=0; i<numArcsToDraw; i++) {
                g.drawArc(start, r, startAngle, sweepAngle);
                r += radiusStep;
            }
            g.drawArc(start, position*2*radius, startAngle, sweepAngle);
        } else {
            // draw the arcs backward from end
            int numArcsToDraw = Math.round(2 * (1f - position) * numArcs);
            float r = numArcs*radiusStep;
            for (int i=0; i<numArcsToDraw; i++) {
                g.drawArc(start, r, startAngle, sweepAngle);
                r -= radiusStep;
            }
            g.drawArc(start, (position-.5f)*2*radius, startAngle, sweepAngle);
        }
    }

    @Override
    protected boolean hidesActor() {
        return false;
    }
}

File: GameLibrary/src/cc/lib/zombicide/anims/SonarAnimation.kt
package cc.lib.zombicide.anims

import cc.lib.game.AGraphics
import cc.lib.game.GColor
import cc.lib.math.Vector2D
import cc.lib.zombicide.ZActor
import cc.lib.zombicide.ZActorAnimation
import cc.lib.zombicide.ZBoard

/**
 * Animate concentric arcs that have a start and end point
 */
class SonarAnimation(actor: ZActor<*>, var start: Vector2D, var end: Vector2D, val numArcs: Int, val sweepAngle: Float) : ZActorAnimation(actor, 2000) {
	var dv: Vector2D
	val startAngle: Float
	val radius: Float

	constructor(actor: ZActor<*>, board: ZBoard, targetZone: Int) : this(actor, actor.getRect().center, board.getZone(targetZone).center, 5, 20f) {}

	override fun draw(g: AGraphics, position: Float, dt: Float) {
		g.color = GColor.WHITE
		g.setLineWidth(3f)
		val radiusStep = radius / numArcs
		if (position <= .5f) {
			// draw the arcs emanating from the start
			val numArcsToDraw = Math.round(position * 2 * numArcs)
			//g.drawFilledCircle(start, radius/10);
			g.drawArc(start, radius / 10, startAngle, sweepAngle)
			var r = radiusStep
			for (i in 0 until numArcsToDraw) {
				g.drawArc(start, r, startAngle, sweepAngle)
				r += radiusStep
			}
			g.drawArc(start, position * 2 * radius, startAngle, sweepAngle)
		} else {
			// draw the arcs backward from end
			val numArcsToDraw = Math.round(2 * (1f - position) * numArcs)
			var r = numArcs * radiusStep
			for (i in 0 until numArcsToDraw) {
				g.drawArc(start, r, startAngle, sweepAngle)
				r -= radiusStep
			}
			g.drawArc(start, (position - .5f) * 2 * radius, startAngle, sweepAngle)
		}
	}

	override fun hidesActor(): Boolean {
		return false
	}

	init {
		radius = end.sub(start).mag()
		startAngle = end.sub(start).angleOf() - sweepAngle / 2
		dv = end.sub(start).scaledBy(1f / numArcs)
	}
}